### PR TITLE
[Fix #3847] Introduce node extensions

### LINKED ISSUE #3847: Introducing node extensions?
One of the first things I [reacted to](https://github.com/bbatsov/rubocop/pull/3012#issue-146355559) when first contributing to RuboCop was how there are utility modules for nodes in the cop mixins. This is a bit unexpected and doesn't make for very discoverable code, and also prone to conflicts with cop methods.

That made me start thinking we should instead have some node extension modules, and decorate the AST nodes with the relevant methods for their respective node types.

Later, I found that [Reek](https://github.com/troessner/reek/tree/master/lib/reek/ast/sexp_extensions) is already doing exactly this. So we could possibly "borrow" some of their code, if it's something you think is worthwhile.

Thoughts? @bbatsov, @jonas054

--------

## Expected behavior

Nodes contain the relevant logic to describe themselves:

```
if_node.ternary?
```

## Actual behavior

Node inspection logic is contained in cop mixins:

```
# Only after mixing in `Cop::Mixin::IfNode`
ternary?(if_node)
```

## RuboCop version

Include the output of `rubocop -V`. Here's an example:

```
$ rubocop -V
0.46.0 (using Parser 2.3.3.1, running on ruby 2.4.0 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by Drenmi**
First of all, sorry for the very large commit. I intended at first, to just do a small part of the node extensions, but whichever end I pulled was attached to some tangled mess down the line, so this change contains a large amount of cop cleanup.

The upside of this is that I has a chance to see whether this approach actually resulted in more understandable code, and I think in a lot of cases, the benefits clearly speak for themselves.

With this approach we can eliminate those long predicate strings of arcane assertions about the node parts, and put them into meaningful methods on the nodes themselves. Anyone with a reasonably powerful text editor can now easily find a method to iterate over the conditions of a `when` node.

Now. On to some things to note about this change:

### Some Rearranging of the AST Related Parts

The `RuboCop` namespace was getting quite crowded, and I wanted to collect all the parts relating to the `AST` in one place, and so I created an `AST` namespace, and moved `Node` (and all its extensions) there. I also changed the file name (to `node.rb`) for parity with the class inside.

### Performance and Approach to Decorating

The first approach I tried was to mix in a module into the eigenclass of each node. Needless to say, this wasn't very performant. (There are a lot (!) of nodes.) It resulted in about a 10% overhead.

After thinking a bit about it, I reverted to basics, and just made subclasses of `Node` with good old inheritance. An `ArrayNode` is, after all, a type of `Node`.

Using this approach rendered the following results:

```
master                         feature/node-extensions
------------------------------------------------------
real      0m51.628s  real      0m51.599s
user      0m50.544s  user      0m50.396s
sys       0m0.655s   sys       0m0.692s
```

* Measured back-to-back on my MacBook Pro (Ruby 2.4.0) with a cold cache and power plugged in.

### Completeness

Not all nodes have their own class yet. There's definitely a lot that can be improved by introducing them. (Some good candidates would be `send` and `pair`.) However, I also did not want to go on forever before making the first PR. 😅 

### Tests and Documentation

I have documented all the node extensions using YARD, and also added test coverage for all of them. I've been diligent about running the whole test suite and RuboCop itself throughout, and I have fairly high confidence in the code.

I have also been rebasing on top of current `master` frequently. (No one likes huge merge conflicts. 😀 )

### Breakage for 3rd Parties

This might possibly break some cops for @backus, if he uses the any of the utility methods found in `Node` that were removed. (There were very few of them.)

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Used the same coding conventions as the rest of the project.
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [X] All tests are passing.
* [X] The new code doesn't generate RuboCop offenses.
* [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/style/conditional_assignment.rb [hunk 1]
```diff
@@ -290,13 +288,11 @@ def autocorrect(node)
           end
         end
 
-        private
-
         def assignment_node(node)
           *_variable, assignment = *node
 
           if assignment.begin_type? && assignment.children.one?
-            assignment, = *assignment
```
Comment (bbatsov) [inline lib/rubocop/cop/style/conditional_assignment.rb]: This was actually intentional for performance reasons (+ I find it more a bit more readable). We discourage the use of `children` across the codebase and rely heavily on destructuring nodes.

Comment (Drenmi) [inline lib/rubocop/cop/style/conditional_assignment.rb]: Noted and fixed. 😀

**Comment by Drenmi** [comment]
@bbatsov: Pushing your suggestion seems to have flushed some cache (?) in Travis, so CI is now passing. 😅

**Comment by Drenmi** [comment]
@bbatsov: Yeah. I'm all too aware of the pain of reviewing big PRs. Sorry about that. 😬

**Comment by backus** [comment]
> This might possibly break some cops for @backus, if he uses the any of the utility methods found in Node that were removed. (There were very few of them.)

Thanks for the heads up.

**Pull request merged**